File: modules/base/types/src/main/java/candle/types/resource/Identifier.java
package candle.types.resource;

import lombok.EqualsAndHashCode;
import lombok.Getter;

import java.util.regex.Pattern;

@Getter
@EqualsAndHashCode
public class Identifier {
  private static final Pattern VALID_NAMESPACE = Pattern.compile("[a-z0-9_.-]+");
  private static final Pattern VALID_PATH = Pattern.compile("[a-z0-9_/.-]+");

  private final String namespace;
  private final String path;

  public Identifier( String namespace, String path ) {
    if ( !isValidNamespace(namespace) ) {
      throw new IllegalArgumentException("Invalid namespace: " + namespace);
    }
    if ( !isValidPath(path) ) {
      throw new IllegalArgumentException("Invalid path: " + path);
    }
    this.namespace = namespace;
    this.path = path;
  }

  // Parses "namespace:path". Defaults to "minecraft" if no colon is present.
  public static Identifier parse( String combined ) {
    String[] parts = combined.split(":", 2);
    if ( parts.length == 1 ) {
      return new Identifier("minecraft", parts[0]);
    }
    return new Identifier(parts[0], parts[1]);
  }

  public static Identifier of( String namespacedIdentifier ) {
    String[] splitIdentifier = namespacedIdentifier.split(":", 1);
    return new Identifier(splitIdentifier[0], splitIdentifier[1]);
  }

  public static Identifier of( String namespace, String identifier ) {
    return new Identifier(namespace, identifier);
  }

  private boolean isValidNamespace( String ns ) {
    return VALID_NAMESPACE.matcher(ns).matches();
  }

  private boolean isValidPath( String path ) {
    return VALID_PATH.matcher(path).matches();
  }

  @Override
  public String toString() {
    return namespace + ":" + path;
  }
}

File: modules/base/types/src/main/kotlin/net/candlemc/types/resource/Identifier.kt
package net.candlemc.types.resource

import java.util.regex.Pattern

open class Identifier(namespace: String, path: String) {
    private val namespace: String
    private val path: String

    init {
        require(isValidNamespace(namespace)) { "Invalid namespace: $namespace" }
        require(isValidPath(path)) { "Invalid path: $path" }
        this.namespace = namespace
        this.path = path
    }

    private fun isValidNamespace(ns: String): Boolean {
        return VALID_NAMESPACE.matcher(ns).matches()
    }

    private fun isValidPath(path: String): Boolean {
        return VALID_PATH.matcher(path).matches()
    }

    override fun toString(): String {
        return "$namespace:$path"
    }

    companion object {
        private val VALID_NAMESPACE: Pattern = Pattern.compile("[a-z0-9_.-]+")
        private val VALID_PATH: Pattern = Pattern.compile("[a-z0-9_/.-]+")

        // Parses "namespace:path". Defaults to "minecraft" if no colon is present.
        fun parse(combined: String): Identifier {
            val parts = combined.split(":".toRegex(), limit = 2).toTypedArray()
            if (parts.size == 1) {
                return Identifier("minecraft", parts[0])
            }
            return Identifier(parts[0], parts[1])
        }

        fun of(namespacedIdentifier: String): Identifier {
            val splitIdentifier = namespacedIdentifier.split(":".toRegex(), limit = 1).toTypedArray()
            return Identifier(splitIdentifier[0], splitIdentifier[1])
        }

        fun of(namespace: String, identifier: String): Identifier {
            return Identifier(namespace, identifier)
        }
    }
}
